### PR TITLE
Sign multi-package XPIs (bug 1160593)

### PR DESCRIPTION
**Pull request by magopian**
Fixes [bug 1160969](https://bugzilla.mozilla.org/show_bug.cgi?id=1160969)

Added two commits in this PR:
1/ detect and store True in `File.is_multi_package`
2/ sign the multi-package XPIs properly (don't sign the XPI itself, sign the internal XPIs if they are extensions)

### CODE REVIEW

File: apps/files/models.py [hunk 1]
```diff
@@ -79,9 +79,12 @@ class File(amo.models.OnChangeMixin, amo.models.ModelBase):
     # file, used for default to compatible.
     binary_components = models.BooleanField(default=False, db_index=True)
     # Serial number of the certificate use for the signature.
-    cert_serial_num = models.CharField(max_length=255, blank=True)
+    cert_serial_num = models.TextField(max_length=255, blank=True)
```
Comment (robhudson) [inline apps/files/models.py]: You no longer meed `max_length`.

File: apps/versions/models.py [hunk 2]
```diff
@@ -490,6 +490,10 @@ def sign_files(self):
         if self.addon.type == amo.ADDON_EXTENSION:
             for file_obj in self.files.all():
                 packaged.sign_file(file_obj)
+        # They can have multi-package XPIs.
+        elif self.addon.type == amo.ADDON_THEME:
+            for file_obj in self.files.filter(is_multi_package=True):
+                packaged.sign_file(file_obj)
```
Comment (robhudson) [inline apps/versions/models.py]: That's the only type?

File: lib/crypto/packaged.py [hunk 3]
```diff
@@ -104,6 +113,34 @@ def sign_file(file_obj):
...
+        with tempfile.NamedTemporaryFile() as temp_file:
+            temp_filename = temp_file.name
+        shutil.make_archive(temp_filename, format='zip', root_dir=folder)
+        # Now overwrite the current multi-package xpi with the repackaged one.
+        # Note that shutil.make_archive automatically added a '.zip' to the end
+        # of the base_name provided as first argument.
+        shutil.move('{0}.zip'.format(temp_filename), file_path)
+        return '\n'.join([str(num) for num in cert_serial_nums])
+    finally:
+        amo.utils.rm_local_tmp_dir(folder)
```
Comment (robhudson) [inline lib/crypto/packaged.py]: Nice work on that.

File: apps/files/models.py (hunk 1 above)
Comment (magopian) [inline apps/files/models.py]: Good catch, thanks ;)

File: apps/versions/models.py (hunk 2 above)
Comment (magopian) [inline apps/versions/models.py]: From what @kmaglione and I have seen, you can only have multi-package XPIs detected as either complete themes or extensions (the code is in https://github.com/mozilla/olympia/blob/75c8700e71de14848bc2d2afbf61362c6c7497c9/apps/files/utils.py#L174).

File: lib/crypto/packaged.py (hunk 3 above)
Comment (magopian) [inline lib/crypto/packaged.py]: It's getting more hairy by the day, and I don't like that, but didn't find a nicer/shorter/clearer way to do it.

**Comment by robhudson** [comment]
I think this looks nice. Well done. You did that faster than I thought. r+wc

**Comment by magopian** [comment]
I must admit it's been a veeery long day.

**Pull request merged**